sw/device/tests/sim_dv/keymgr_key_derivation.c

```diff
@@ -0,0 +1,254 @@
+    0xb4fb0fdf, 0x1ca1567b, 0x13a0e876, 0xa6521d8f};
+
+/**
+ * Key manager Owner Secret stored in info flash page.
+ */
+const uint32_t kOwnerSecret[kSecretWordSize] = {
+    0xa6521d8f, 0x13a0e876, 0x1ca1567b, 0xb4fb0fdf,
+    0x9f89bc56, 0x4bd127c7, 0x322288d8, 0x4e919d54,
+};
+
+/**
+ * Kmac prefix "KMAC" with empty custom string
+ */
+#define KMAC_PREFIX_SIZE 11
+const uint32_t kKmacPrefix[KMAC_PREFIX_SIZE] = {
+    0x4d4b2001, 0x00014341, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+};
+
+/** ROM_EXT key manager maximum version. */
+const uint32_t kMaxVerRomExt = 1;
+
+/** BL0 key manager maximum version. */
+const uint32_t kMaxVerBl0 = 2;
+
+const test_config_t kTestConfig;
+
+/**
+ * Writes `size` words of `data` into flash info page.
+ *
+ * @param page_id Info page ID to write to.
+ * @param data Data to write.
+ * @param size Number of 4B words to write from `data` buffer.
+ */
+static void write_info_page(uint32_t page_id, const uint32_t *data,
+                            size_t size) {
+  mp_region_t info_region = {.num = page_id,
+                             .base = 0x0,  // only used to calculate bank id.
+                             .size = 0x1,  // unused for info pages.
+                             .part = kInfoPartition,
+                             .rd_en = true,
+                             .prog_en = true,
+                             .erase_en = true,
+                             .scramble_en = false};
+  flash_cfg_region(&info_region);
+
+  uint32_t address = FLASH_MEM_BASE_ADDR + page_id * flash_get_page_size();
+
+  ASSERT_EQZ(flash_page_erase(address, kInfoPartition));
+  ASSERT_EQZ(flash_write(address, kInfoPartition, data, size));
+
+  for (size_t i = 0; i < size; ++i) {
+    uint32_t got_data;
+    ASSERT_EQZ(flash_read(address + i * sizeof(uint32_t), kInfoPartition,
+                          /*size=*/1, &got_data));
+    CHECK(got_data == data[i]);
+  }
+  info_region.rd_en = false;
+  info_region.prog_en = false;
+  info_region.erase_en = false;
+  flash_cfg_region(&info_region);
+}
+
+static void init_flash(void) {
+  // setup default access for data partition
+  flash_default_region_access(/*rd_en=*/true, /*prog_en=*/true,
+                              /*erase_en=*/true);
+  // Initialize flash secrets.
+  write_info_page(kFlashInfoPageIdCreatorSecret, kCreatorSecret,
+                  ARRAYSIZE(kCreatorSecret));
+  write_info_page(kFlashInfoPageIdOwnerSecret, kOwnerSecret,
+                  ARRAYSIZE(kOwnerSecret));
+}
+
+/**
+ * Busy-wait until the DAI is done with whatever operation it is doing.
+ */
+static dif_otp_ctrl_t otp;
+
+/**
+ * Lock otp secret 2 partition so that on reboot flash seeds can be
+ * automatically loaded
+ */
+static void lock_otp_secret_partition2(void) {
+  // initialize otp
+  CHECK_DIF_OK(dif_otp_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR), &otp));
+
+  otp_ctrl_testutils_lock_partition(&otp, kDifOtpCtrlPartitionSecret2, 0);
+}
```

Comment:
I think you can just use the testutil directly, and do the `dif_otp_ctrl_init()` in the test_main somewhere. Then you can get rid of this, and just use the testutil directly.

Reply:
oops, looks like @sriyerg said the same below. sorry for the double comment :)

Reply:
> Thanks @weicaiyang, looks much better with the DIFs. With the exception of one comment on extra OTP lock function, this LGTM from a SW side. One more improvement would be to use the actual flash_ctrl DIFs instead of the legacy DIF-like code. But I'm ok with that being addressed in a future PR. Perhaps @a-will can help with that.
> 
> On a separate note, now that the flash_ctrl DIFs have been completed, we should probably remove the `sw/device/lib/flash_ctrl.{c,h}` files from the repo and switch all existing code to use the DIFs instead.

I just addressed all these in my last push. haha. PTAL